panda/src/navigation/navMesh.cxx

```diff
@@ -0,0 +1,388 @@
+  }
+
+  _nav_mesh = dtAllocNavMesh();
+
+  if (!_nav_mesh) {
+    dtFree(nav_data);
+    std::cout<<"\nCould not create Detour navmesh\n";
+    return false;
+  }
+
+  dtStatus status;
+
+  status = _nav_mesh->init(nav_data, nav_data_size, DT_TILE_FREE_DATA);
+  if (dtStatusFailed(status)) {
+    dtFree(nav_data);
+    std::cout<<"\nCould not init Detour navmesh\n";
+    return false;
+  }
+}
+
+
+PT(GeomNode) NavMesh::draw_nav_mesh_geom() {
+
+  PT(GeomVertexData) vdata;
+  vdata = new GeomVertexData("vertexInfo", GeomVertexFormat::get_v3c4(), Geom::UH_static);
+  std::cout << "\nJust degbugging for navMesh: navmesh vert count: " << _params.vertCount << "\n";
+  vdata->set_num_rows(_params.vertCount);
+
+  GeomVertexWriter vertex(vdata, "vertex");
+  GeomVertexWriter colour(vdata, "color");
+
+  const int nvp = _params.nvp;
+  std::cout << "nvp: " << nvp << std::endl;
+  const float cs = _params.cs;
+  std::cout << "cs: " << cs << std::endl;
+  const float ch = _params.ch;
+  std::cout << "ch: " << ch << std::endl;
+  const float* orig = _params.bmin;
+  std::cout << "orig: " << orig[0] << "\t" << orig[1] << "\t" << orig[2] << std::endl;
+
+  std::cout << "NavMesh poly count: " << _params.polyCount << std::endl;
+  std::cout << "NavMesh vert count: " << _params.vertCount << std::endl;
+
+  for (int i = 0;i < _params.vertCount * 3;i += 3) {
+
+    const unsigned short* v = &_params.verts[i];
+
+    //convert to world space
+    const float x = orig[0] + v[0] * cs;
+    const float y = orig[1] + v[1] * ch;
+    const float z = orig[2] + v[2] * cs;
+
+    vertex.add_data3(x, -z, y); //if origingally model is z-up
+    //vertex.add_data3(x, y, z); //if originally model is y-up
+    colour.add_data4((float)rand() / RAND_MAX, (float)rand() / RAND_MAX, (float)rand() / RAND_MAX, 1);
+    //std::cout << "index: " << i / 3 << "\t" << x << "\t" << y << "\t" << z << "\n";
+
+  }
+
+  PT(GeomNode) node;
+  node = new GeomNode("gnode");
+
+  PT(GeomTrifans) prim;
+  prim = new GeomTrifans(Geom::UH_static);
+
+  for (int i = 0; i < _params.polyCount; ++i) {
+
+
+    const unsigned short* p = &_params.polys[i*nvp * 2];
+
+    // Iterate the vertices.
+    //unsigned short vi[3];  // The vertex indices.
+    for (int j = 0; j < nvp; ++j) {
+      if (p[j] == RC_MESH_NULL_IDX) {
+        break;// End of vertices.
+      }
+      if (p[j + nvp] == RC_MESH_NULL_IDX) {
+        prim->add_vertex(p[j]);
+        // The edge beginning with this vertex is a solid border.
+      }
+      else {
+        prim->add_vertex(p[j]);
+        // The edge beginning with this vertex connects to 
+        // polygon p[j + nvp].
+      }
+      //std::cout << "p[j]: " << p[j] << std::endl;
+
+    }
+    prim->close_primitive();
+
+  }
+  PT(Geom) polymeshgeom;
+  polymeshgeom = new Geom(vdata);
+  polymeshgeom->add_primitive(prim);
+
+  node->add_geom(polymeshgeom);
+  std::cout << "Number of Polygons: " << _params.polyCount << std::endl;
+
+
+  return node;
+}
+
+
+
+
+
```

Comment:
Please remove unnecessary lines of spacing.  One empty line between methods is enough.

Reply:
Done